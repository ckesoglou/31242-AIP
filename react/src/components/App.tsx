import React from 'react';
import '../assets/css/App.css';
import {
  Switch,
  Route,
  Link
} from "react-router-dom";
import Login from '../pages/login';
import SignUp from '../pages/signup';

function App() {
  return (
    <div>
      <Link to="/about">About</Link>
      <Link to="/users">User</Link>
      <Link to="/home">Home</Link>
      <Link to="/">Login</Link>
      <Link to="/signup">Sign Up</Link>

      <Switch>
        <Route path="/about">
          <h1>About!</h1>
        </Route>
        <Route path="/users">
          <h1>Users!</h1>
        </Route>
        <Route path="/home">
          <h1>Home!</h1>
        </Route>
        <Route path="/signup">
          <SignUp />
        </Route>
        <Route path="/">
          <Login />
        </Route>
      </Switch>
    </div>
  );
}

export default App;
