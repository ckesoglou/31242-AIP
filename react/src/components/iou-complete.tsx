import React from "react";
import {
  Checkbox,
  Dialog,
  Grow,
  DialogTitle,
  Typography,
  DialogContent,
  DialogContentText,
  DialogActions,
  Button,
  Snackbar,
  Divider,
  TableContainer,
  Table,
  TableBody,
  TableRow,
  TableCell,
  Popover,
} from "@material-ui/core";
import CheckCircleOutlineIcon from "@material-ui/icons/CheckCircleOutline";
import RadioButtonUncheckedIcon from "@material-ui/icons/RadioButtonUnchecked";
import "../assets/css/iou-request.css";
import IouFavour from "./iou-single-favour";
import { imageEndpoint, requestsEndpoint } from "../api/endpoints";

type Item = {
  id: string;
  display_name: string;
};

type RequestObj = {
  id: string;
  author: { username: string; display_name: string };
  completed_by: { username: string; display_name: string };
  proof_of_completion: string;
  rewards: Item[];
  details: string;
  created_time: string;
  comletion_time: string;
  is_completed: boolean;
};

type IouCompleteProps = {
  request: RequestObj;
};

type IouCompleteState = {
  completeIOU: boolean;
  submittedProof: any;
  snackMessage: string;
  completeSnack: boolean;
  AnchorEl: HTMLElement | null;
};

class IouComplete extends React.Component<IouCompleteProps, IouCompleteState> {
  private tempRewardDisplayName: string;

  constructor(props: IouCompleteProps) {
    super(props);

    this.tempRewardDisplayName = "";
    this.openCompleteForm = this.openCompleteForm.bind(this);
  }

  state: IouCompleteState = {
    completeIOU: false,
    submittedProof: null,
    snackMessage: "",
    completeSnack: false,
    AnchorEl: null,
  };

  openCompleteForm() {
    if (!this.props.request.is_completed) {
      this.setState({ completeIOU: true });
    }
  }

  submitProof() {
    fetch(`${imageEndpoint}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        // requestID: ???,
        // proof: this.state.submittedProof ???
      }),
    })
      .then((res) => {
        return res.json();
      })
      .then((body) => {
        console.log("Success:", body);
        this.setState({ snackMessage: "IOU Completion proof submitted!" });
        this.setState({ completeSnack: true });
      })
      .catch((exception) => {
        console.error("Error:", exception);
        this.setState({ snackMessage: exception });
        this.setState({ completeSnack: true });
      });
  }

  fileContent() {
    if (this.state.submittedProof) {
      return (
        <div id="completeProofFileInfo">
          <DialogContentText variant="body2">
            {"Image Preview: "}
          </DialogContentText>
          <img
            src={URL.createObjectURL(this.state.submittedProof)}
            alt={this.state.submittedProof.name}
            id="completeProofImage"
          />
        </div>
        //Need to discuss how submitted images should be formatted (Size, Encode Format)
      );
    } else {
      return (
        <DialogContentText variant="body2" id="completeProofFileInfo">
          {"Upload an image (JPEG/PNG) before pressing the 'Submit' button"}
        </DialogContentText>
      );
    }
  }

  fetchRewardDetails(requestID: string, rewardID: string) {
    fetch(
      `${requestsEndpoint
        .concat(requestID)
        .concat("/reward/")
        .concat(rewardID)}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      }
    )
      .then((res) => {
        return res.json();
      })
      .then((body) => {
        console.log("Success:", body);
        this.tempRewardDisplayName = body.giver.display_name;
      })
      .catch((exception) => {
        console.error("Error:", exception);
        this.tempRewardDisplayName = "Could not find user display name";
      });
  }

  renderPopUpRewards(item: Item) {
    //this.fetchRewardDetails(this.props.request.id, item.id);
    this.tempRewardDisplayName = "James Lee";
    return (
      <IouFavour
        key={item.id}
        giverDisplayName={this.tempRewardDisplayName}
        recieverDisplayName={"?"}
        item={item}
      />
    );
  }

  render() {
    return (
      <div id="completeRequestItem">
        <div
          onClick={() => {
            if (!this.props.request.is_completed) {
              this.setState({ completeIOU: true });
            }
          }}
          className={this.props.request.is_completed ? "" : "cursorPointer"}
        >
          <Checkbox
            checked={this.props.request.is_completed}
            color="primary"
            disabled={true}
            checkedIcon={
              <CheckCircleOutlineIcon color="primary" id="completeIconSize" />
            }
            icon={
              <RadioButtonUncheckedIcon color="primary" id="completeIconSize" />
            }
          />
          {this.props.request.is_completed && (
            <div id="IouTaskComplete">
              {this.props.request.completed_by.display_name.length > 9 ? (
                <Typography
                  id="taskCompleter"
                  className="cursorPointer"
                  onClick={(event: React.MouseEvent<HTMLElement>) =>
                    this.setState({ AnchorEl: event.currentTarget })
                  }
                >
                  {this.props.request.completed_by.display_name}
                </Typography>
              ) : (
                <Typography id="taskCompleter">
                  {this.props.request.completed_by.display_name}
                </Typography>
              )}
              <Popover
                id="popUpMargin"
                open={Boolean(this.state.AnchorEl)}
                anchorEl={this.state.AnchorEl}
                onClose={() => this.setState({ AnchorEl: null })}
                anchorOrigin={{
                  vertical: "bottom",
                  horizontal: "center",
                }}
                transformOrigin={{
                  vertical: "top",
                  horizontal: "center",
                }}
              >
                <Typography id="taskDetailPopUp">
                  {"Full Display Name"}
                  <Divider id="taskPopUpDivider" />
                  {this.props.request.completed_by.display_name}
                </Typography>
              </Popover>
              <Typography id="taskTimeStamp">
                {this.props.request.comletion_time}
              </Typography>
            </div>
          )}
        </div>
        <Dialog
          maxWidth="xs"
          fullWidth={true}
          scroll="body"
          open={this.state.completeIOU}
          TransitionComponent={Grow}
          onClose={() =>
            this.setState({ completeIOU: false, submittedProof: null })
          }
          id="completeForm"
        >
          <DialogTitle disableTypography={true}>
            <Typography variant="h5" id="modalTitle">
              {"IOU Proof Submission"}
            </Typography>
          </DialogTitle>
          <Divider />
          <DialogContent className="content" id="completeDetailsTable">
            <DialogContentText variant="h6">{"IOU Details"}</DialogContentText>
            <TableContainer id="infoTable">
              <Table>
                <TableBody>
                  <TableRow>
                    <TableCell
                      align="center"
                      className="infoTableCell"
                      id="infoTableTitle"
                    >
                      Created by:
                    </TableCell>
                    <TableCell
                      align="center"
                      className="infoTableCell"
                      id="infoTableContent"
                    >
                      {this.props.request.author.display_name}
                    </TableCell>
                  </TableRow>
                  <TableRow>
                    <TableCell align="center" id="infoTableTitle">
                      Created on:
                    </TableCell>
                    <TableCell align="center" id="infoTableContent">
                      {this.props.request.created_time}
                    </TableCell>
                  </TableRow>
                  <TableRow>
                    <TableCell align="center" id="infoTableTitle">
                      Task:
                    </TableCell>
                    <TableCell align="center" id="infoTableContent">
                      {this.props.request.details}
                    </TableCell>
                  </TableRow>
                </TableBody>
              </Table>
            </TableContainer>
          </DialogContent>
          <DialogContent className="Content" id="completePopUpRewards">
            <DialogContentText variant="h6">{"Rewards"}</DialogContentText>
            {this.props.request.rewards.map((item) =>
              this.renderPopUpRewards(item)
            )}
          </DialogContent>
          <DialogContent className="content">
            <DialogContentText variant="h6">
              {"Upload proof?"}
            </DialogContentText>
            <input
              type="file"
              onChange={(e) => {
                if (e.target.files) {
                  this.setState({ submittedProof: e.target.files[0] });
                }
              }}
              accept="image/*"
              id="inputProof"
            />
            <label htmlFor="inputProof">
              <Button variant="contained" color="primary" component="span">
                {this.state.submittedProof ? "CHANGE" : "UPLOAD"}
              </Button>
            </label>
            {this.fileContent()}
          </DialogContent>
          <Divider />
          <DialogActions>
            <Button
              id="createRequest"
              size="large"
              color="primary"
              onClick={() => {
                this.submitProof();
                this.setState({
                  completeIOU: false,
                });
              }}
              autoFocus
              disabled={!this.state.submittedProof}
            >
              Submit
            </Button>
            <Button
              size="large"
              color="primary"
              onClick={() =>
                this.setState({ completeIOU: false, submittedProof: null })
              }
            >
              Nevermind
            </Button>
          </DialogActions>
        </Dialog>
        <Snackbar
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "left",
          }}
          message={this.state.snackMessage}
          open={this.state.completeSnack}
          onClose={() => {
            this.setState({ completeSnack: false });
          }}
          autoHideDuration={5000}
        />
      </div>
    );
  }
}

export default IouComplete;
