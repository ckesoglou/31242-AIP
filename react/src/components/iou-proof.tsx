import React from "react";
import "../assets/css/iou-request.css";
import { imageEndpoint } from "../api/endpoints";
import ClearIcon from "@material-ui/icons/Clear";
import ImageIcon from "@material-ui/icons/Image";
import { Popover, Typography, Divider } from "@material-ui/core";

type IouProofProps = {
  imagePK: string;
};

type IouProofState = {
  resImage: string;
  anchorEl: HTMLElement | null;
};

class IouProof extends React.Component<IouProofProps, IouProofState> {
  state: IouProofState = {
    resImage:
      "iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAYAAAAeP4ixAAAABGdBTUEAAK/INwWK6QAAAAlwSFlzAAAOxAAADsQBlSsOGwAAABl0RVh0U29mdHdhcmUAQWRvYmUgSW1hZ2VSZWFkeXHJZTwAAAUXSURBVGhD3Zk/iBRXHMffmCJrp0UOTkJgDSTZMlcImiKgsVHBsNeEbBHwjljcFkIEE9KkyR8hgRR7heJZCBfSuCQQ0wQNqY6kuCsnEXTBwgNTaOdaBPM+b++399vZNzM7b2ZX9AfDnjPvz+/z+/7ee78Zo6fWTEV2+fa/QSN9/MYrQf10pz2lR1ADVOFQqD+VgoQ4URX8MwcJgff1qRQkdI1UARPlLXbtXF4ahILkjZsEXVuMhreWrg/2qujSPw+epg2U5pivfSiEeJQ1pn4GRKvTHIKst7sGGAdShbTTHkNgequRmW/ugjAvMJWukWnCoLgPQuZ8bkCygkSqPTcgx3+bG0upUif72TfnzPfvvzvNLAoaO1gRYGYFlKcG5MEgErZZweTJVBqECV76bC5vnuDnk6hRiSKbnX6wk1V2LKyIPUBNHMfOh0bjbffbWqlPRZVJ1SilCBDrKxtVBnVkLNbeydWGWTgWu8vEW6lz7T3QDStR2IIf36yPDXzt1T9LgbHWvosbI2rLgLEF+aQRO/UHDQbZAARZElxr/fXBa6bz7fyI4yEgeqPghG61a95gAKLX495jPdcOCKxSkPXVnjl6dNePW7dGffrv68GkmADoSlaeJWF8EAIgfYJBGECrsh33TH1pfSya/YfXh/eoUofOqlI82UlAAGBj0WncPr9tDv14b2yeUiASWYlqbf+iNy00jGvAwt3JcV8H8l7siq3Yh+ti56YPphQI6UEqyfvBxCAe77XzybRhc8GSyqxt9cuvkaQa/NsHMqaGgtjuds2XazWjHfJKunPTt1uy6EvtWqiB5AzkIpWRKlnOkWbJHcjXPguC9oVPdj0JDvChUhzxOoyjNudd6vgONRsAAsFawFl9yXh5EKVBGGBxsZUZ8NbqYQeLLaScETxD3bsd44C4+FugeM6uiBE0SSc9cbAiG72aeXy/abfHLTdxWgnR7f5goigyzeaHmcA8nG/UhzsUZ5AoxfgH2wMYgdWQ9A3ataiDNm8OSglt7ZO/ms+X+qOvpGoNeNeSJ91wGBC3I1o4OWiBcUFTtmwzlsUerIgvvJ0bJ8bfq1kDVjkuNoSH5x+5K8twHtOVQbJK4LlA8HcwiHNmJ5oogfGrT2+5J05rgM13bOWcUdHqUkcORBTFeZ8FgZz76Q/T24jN5rJ9qbLOHK73HQRFJJMCw8U5gUrObDv6yJUpiX0oqqS102rQJmiN0JE668yj/Wbhir9a9TngwJUV6cu6EThKlOQhuufi32fyguN9TuG2XNtXqK92vAgEk2iFfJVAdPDnU8NvvxfeulrIMRrLV5RZvrtPfI4UVYk3N7ZI5JeDq3BESnYYUSQ5FgoBlaWU1FzSFyAx2XnyFq7O/0l4tCLiXyaID4x70tkdjBV8DioCIhDJrCm0/dJZD0BKVWG+wy5rXF/qFwLRg6OGezP0lO86vaRPFWuHbdduTl7GYJCR0YBRQHzSodLVQC7q0i7RngI0zwgE226aBYMcqe0uaj04AE++ed1dABFFLq+zfOSzmwMOkvu08xkQFIxpatCn0GKXSV7+9I6rfilBZEfCIRwHQBttb6zE7quh+8CW+HpCafP7F+8Nu9w9/Yu3ws2CoHOQIjjLZ0zSBQW4vvro1BgEE9AWCMypQqEol73nSxcpEvlNKxKTygUpkpfPac+JNuW8U9L+zyyvv75I004sTwlpN1MQJg1xcpLAzRxkEqdC2gStkZCJpt3nhQH5H9ulgVHYe0gsAAAAAElFTkSuQmCC",
    anchorEl: null,
  };

  fetchImage() {
    fetch(`${imageEndpoint + this.props.imagePK}`, {
      method: "GET",
    })
      .then((res) => {
        return res.json();
      })
      .then((body) => {
        this.setState({
          resImage: body,
        });
      });
  }

  render() {
    return (
      <div id="requestItem">
        {this.props.imagePK ? (
          <div>
            <div
              className="cursorPointer"
              onClick={(event: React.MouseEvent<HTMLElement>) => {
                this.setState({ anchorEl: event.currentTarget });
              }}
            >
              <ImageIcon color="primary" id="proofIcon" />
            </div>
            <Popover
              id="popUpMargin"
              open={Boolean(this.state.anchorEl)}
              anchorEl={this.state.anchorEl}
              onClose={() => this.setState({ anchorEl: null })}
              anchorOrigin={{
                vertical: "bottom",
                horizontal: "center",
              }}
              transformOrigin={{
                vertical: "top",
                horizontal: "center",
              }}
            >
              <Typography id="taskDetailPopUp">
                {"Proof"}
                <Divider id="taskPopUpDivider" />
                <img
                  src={"data:image/jpeg;base64," + this.state.resImage}
                  id="proofImage"
                  alt="Proof of Completion"
                />
              </Typography>
            </Popover>
          </div>
        ) : (
          <ClearIcon id="noProofIcon" />
        )}
      </div>
    );
  }
}

export default IouProof;
