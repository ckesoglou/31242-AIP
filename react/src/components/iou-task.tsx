import React from "react";
import "../assets/css/iou-request.css";
import { Popover, Typography, Divider } from "@material-ui/core";

type IouTaskProps = {
  details: string;
};

type IouTaskState = {
  anchorEl: HTMLElement | null;
};

class IouTask extends React.Component<IouTaskProps, IouTaskState> {
  state: IouTaskState = {
    anchorEl: null,
  };

  render() {
    return (
      <div
        className={this.props.details.length > 55 ? "cursorPointer" : ""}
        id="task"
      >
        {this.props.details.length > 55 ? (
          <div
            id="taskDetailShort"
            onClick={(event: React.MouseEvent<HTMLElement>) => {
              this.setState({
                anchorEl: event.currentTarget,
              });
            }}
          >
            {this.props.details
              .substring(
                0,
                this.props.details.length - (this.props.details.length - 55) - 3
              )
              .concat("...")}
          </div>
        ) : (
          <Typography>{this.props.details}</Typography>
        )}
        <Popover
          id="popUpMargin"
          open={Boolean(this.state.anchorEl)}
          anchorEl={this.state.anchorEl}
          onClose={() => this.setState({ anchorEl: null })}
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "center",
          }}
          transformOrigin={{
            vertical: "top",
            horizontal: "center",
          }}
        >
          <Typography id="taskDetailPopUp">
            {"Task Description"}
            <Divider id="taskPopUpDivider" />
            {this.props.details}
          </Typography>
        </Popover>
      </div>
    );
  }
}

export default IouTask;
