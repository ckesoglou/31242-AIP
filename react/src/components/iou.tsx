import React from "react";
import IouComplete from "./iou-complete";
import IouProof from "./iou-proof";
import IouFavour from "./iou-single-favour";
import RequestInfo from "./request-info";
import { Grid } from "@material-ui/core";

type Item = {
  id: string;
  display_name: string;
};

type RequestObj = {
  id: string;
  author: { username: string; display_name: string };
  completed_by: { username: string; display_name: string };
  proof_of_completion: string;
  rewards: Item[];
  details: string;
  created_time: string;
  comletion_time: string;
  is_completed: boolean;
};

type RequestProps = {
  request: RequestObj;
};

class IOU extends React.Component<RequestProps> {
  render() {
    return (
      <div>
        <Grid container xs={12} spacing={1}>
          <Grid item xs={6} id="requestItemContainer">
            <IouFavour
              giverDisplayName={this.props.request.author.display_name}
              recieverDisplayName={"?"}
              item={this.props.request.rewards[0]}
            />
          </Grid>
          <Grid item xs={1} id="requestItemContainer">
            <IouProof imagePK={this.props.request.proof_of_completion} />
          </Grid>
          <Grid item xs={3} id="requestProofContainer">
            <IouComplete request={this.props.request} />
          </Grid>
          <Grid item xs={1} id="requestItemContainer">
            <RequestInfo request={this.props.request} />
          </Grid>
        </Grid>
      </div>
    );
  }
}

export default IOU;
