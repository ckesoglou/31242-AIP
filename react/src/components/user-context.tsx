import React from "react";

const userContext = React.createContext({
  user: {},
  updateUser: (newUser: object) => {},
});

export { userContext as UserContext };
