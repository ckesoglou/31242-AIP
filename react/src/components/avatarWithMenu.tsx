import React from "react";
import { Link as RouterLink } from "react-router-dom";
import "../assets/css/avatarWithMenu.css";
import {
  Avatar,
  Link,
  ListItemIcon,
  Menu,
  MenuItem,
  MenuProps,
  withStyles,
  ListItemText,
} from "@material-ui/core";
import {
  PeopleOutline,
  ThumbsUpDown,
  ExitToApp,
  AssignmentInd,
  PersonAdd,
  AccountCircle,
} from "@material-ui/icons";
import { UserContext } from "./user-context";

// kudos to https://material-ui.com/components/menus/
const StyledMenu = withStyles({
  paper: {
    border: "1px solid #d3d4d5",
  },
})((props: MenuProps) => (
  <Menu
    elevation={0}
    getContentAnchorEl={null}
    anchorOrigin={{
      vertical: "bottom",
      horizontal: "center",
    }}
    transformOrigin={{
      vertical: "top",
      horizontal: "center",
    }}
    {...props}
  />
));

// kudos to https://material-ui.com/components/menus/
const StyledMenuItem = withStyles((theme) => ({
  root: {
    "&:focus": {
      backgroundColor: theme.palette.primary.main,
      "& .MuiListItemIcon-root, & .MuiListItemText-primary": {
        color: theme.palette.common.white,
      },
    },
  },
}))(MenuItem);

type AvatarWithMenuProps = {
  fullName: string;
  loggedIn: boolean;
};

type AvatarWithMenuState = {
  userMenu: boolean;
  anchorEl: HTMLElement | null;
};

// most readable solution from https://stackoverflow.com/questions/33076177/getting-name-initials-using-js
function nameToUpperInitials(fullName: string) {
  const namesArray = fullName.split(" ");
  if (namesArray.length === 1)
    return `${namesArray[0].charAt(0).toUpperCase()}`;
  else
    return `${namesArray[0].charAt(0).toUpperCase()}${namesArray[
      namesArray.length - 1
    ]
      .charAt(0)
      .toUpperCase()}`;
}

class AvatarWithMenu extends React.Component<
  AvatarWithMenuProps,
  AvatarWithMenuState
> {
  state: AvatarWithMenuState = {
    userMenu: false,
    anchorEl: null,
  };

  static contextType: React.Context<{
    user: {};
    updateUser: (newUser: object) => void;
  }> = UserContext;

  render() {
    return (
      <div>
        {this.props.loggedIn && (
          <div>
            <Avatar
              onClick={(event: React.MouseEvent<HTMLElement>) => {
                this.setState({
                  userMenu: true,
                  anchorEl: event.currentTarget,
                });
              }}
              id="avatar"
            >
              {nameToUpperInitials(this.props.fullName)}
            </Avatar>
            <StyledMenu
              id="customized-menu"
              anchorEl={this.state.anchorEl}
              keepMounted
              open={this.state.userMenu}
              onClose={() => {
                this.setState({
                  userMenu: false,
                  anchorEl: null,
                });
              }}
            >
              <Link
                id="favoursLink"
                color="inherit"
                component={RouterLink}
                to={{ pathname: "/user", state: { tabIndex: 1 } }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <ThumbsUpDown fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="My Favours" />
                </StyledMenuItem>
              </Link>
              <Link
                id="requestLink"
                color="inherit"
                component={RouterLink}
                to={{ pathname: "/user", state: { tabIndex: 2 } }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <PeopleOutline fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="My Requests" />
                </StyledMenuItem>
              </Link>
              <MenuItem />
              <Link
                id="logOutLink"
                color="inherit"
                component={RouterLink}
                to="/home"
                onClick={() => {
                  this.context.updateUser({ name: "?" });
                }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <ExitToApp fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="Log Out" />
                </StyledMenuItem>
              </Link>
            </StyledMenu>
          </div>
        )}
        {!this.props.loggedIn && (
          <div>
            <Avatar
              onClick={(event: React.MouseEvent<HTMLElement>) => {
                this.setState({
                  userMenu: true,
                  anchorEl: event.currentTarget,
                });
              }}
              id="avatar"
            >
              <AccountCircle id="accountCircle" />
            </Avatar>
            <StyledMenu
              id="customized-menu"
              anchorEl={this.state.anchorEl}
              keepMounted
              open={this.state.userMenu}
              onClose={() => {
                this.setState({
                  userMenu: false,
                  anchorEl: null,
                });
              }}
            >
              <Link
                id="favoursLink"
                color="inherit"
                component={RouterLink}
                to={{ pathname: "/login", state: { tabIndex: 1 } }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <AssignmentInd fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="Login" />
                </StyledMenuItem>
              </Link>
              <Link
                id="requestLink"
                color="inherit"
                component={RouterLink}
                to={{ pathname: "/signup", state: { tabIndex: 2 } }}
                style={{ textDecoration: "none" }}
              >
                <StyledMenuItem>
                  <ListItemIcon>
                    <PersonAdd fontSize="small" />
                  </ListItemIcon>
                  <ListItemText primary="Sign Up" />
                </StyledMenuItem>
              </Link>
            </StyledMenu>
          </div>
        )}
      </div>
    );
  }
}

export { AvatarWithMenu, nameToUpperInitials };
