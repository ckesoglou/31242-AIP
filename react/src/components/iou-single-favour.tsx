import React from "react";
import "../assets/css/iou-request.css";
import { KeyboardArrowRight as Connector } from "@material-ui/icons";
import { Avatar, Popover, Typography } from "@material-ui/core";
import { nameToUpperInitials } from "./avatarWithMenu";

type Item = {
  id: string;
  display_name: string;
};

type IouFavourSingleProps = {
  giverDisplayName: string;
  recieverDisplayName: string;
  item: Item;
};

type IouFavourSingleState = {
  giverAnchorEl: HTMLElement | null;
  recieverAnchorEl: HTMLElement | null;
};

class IouFavourSingle extends React.Component<
  IouFavourSingleProps,
  IouFavourSingleState
> {
  state: IouFavourSingleState = {
    giverAnchorEl: null,
    recieverAnchorEl: null,
  };

  render() {
    return (
      <div id="requestItem">
        <div id="favourAvatarContainer">
          <Avatar
            onClick={(event: React.MouseEvent<HTMLElement>) => {
              this.setState({
                giverAnchorEl: event.currentTarget,
              });
            }}
            id="favourAvatar"
          >
            {nameToUpperInitials(this.props.giverDisplayName)}
          </Avatar>
          <Popover
            id="popUpMargin"
            open={Boolean(this.state.giverAnchorEl)}
            anchorEl={this.state.giverAnchorEl}
            onClose={() => this.setState({ giverAnchorEl: null })}
            anchorOrigin={{
              vertical: "bottom",
              horizontal: "center",
            }}
            transformOrigin={{
              vertical: "top",
              horizontal: "center",
            }}
          >
            <Typography id="favourAvatarText">
              {this.props.giverDisplayName}
            </Typography>
          </Popover>
        </div>
        <Connector color="primary" id="favourArrow" />
        <div id="rewardContainer">
          <Avatar id="rewardAvatar">{this.props.item.display_name}</Avatar>
        </div>
        <Connector color="primary" id="favourArrow" />
        <div id="favourAvatarContainer">
          <Avatar
            onClick={(event: React.MouseEvent<HTMLElement>) => {
              this.setState({
                recieverAnchorEl: event.currentTarget,
              });
            }}
            id="favourAvatar"
          >
            {nameToUpperInitials(this.props.recieverDisplayName)}
          </Avatar>
          <Popover
            id="popUpMargin"
            open={Boolean(this.state.recieverAnchorEl)}
            anchorEl={this.state.recieverAnchorEl}
            onClose={() => this.setState({ recieverAnchorEl: null })}
            anchorOrigin={{
              vertical: "bottom",
              horizontal: "center",
            }}
            transformOrigin={{
              vertical: "top",
              horizontal: "center",
            }}
          >
            <Typography id="favourAvatarText">
              {this.props.recieverDisplayName === "?"
                ? "Favour not completed by anyone yet!"
                : this.props.recieverDisplayName}
            </Typography>
          </Popover>
        </div>
      </div>
    );
  }
}

export default IouFavourSingle;
