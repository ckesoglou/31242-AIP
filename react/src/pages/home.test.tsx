import React from "react";
import { shallow, mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import Home from "./home";
import UserProfile from "./userprofile";
import { UserContext } from "../components/user-context";

const testProps = {
  history: {} as any,
  location: {} as any,
  match: {} as any,
};

describe("<Home />", () => {
  it("should render correctly", () => {
    const wrapper = shallow(
      <UserContext.Provider
        value={{
          user: { name: "Kevin Leung" },
          updateUser: (newUser: Object) => {},
        }}
      >
        <Home />
      </UserContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("navigates to user page from user menu", () => {
    const wrapper = mount(
      <UserContext.Provider
        value={{
          user: { name: "Kevin Leung" },
          updateUser: (newUser: Object) => {},
        }}
      >
        <MemoryRouter initialEntries={["/home"]}>
          <Route path="/home" component={Home} />
          <Route path="/user" component={UserProfile} />
        </MemoryRouter>
      </UserContext.Provider>
    );
    let homeComponent = wrapper.find("Home");
    homeComponent.find("div#avatar").simulate("click");
    // kudos to https://github.com/enzymejs/enzyme/issues/516
    homeComponent.find("a#favoursLink").simulate("click", { button: 0 });

    // @ts-ignore
    expect(wrapper.find("Router").prop("history").location.pathname).toEqual(
      "/user"
    );
    // @ts-ignore
    expect(wrapper.find("UserProfile").prop("location").state.tabIndex).toEqual(
      1
    );
  });

  it("should have initials of logged in user for avatar", () => {
    const spy = jest.spyOn(Home.prototype, "nameToUpperInitials");
    const wrapper = mount(
      <UserContext.Provider
        value={{
          user: { name: "Kevin Leung" },
          updateUser: (newUser: Object) => {},
        }}
      >
        <MemoryRouter>
          <Home />
        </MemoryRouter>
      </UserContext.Provider>
    );

    let avatar = wrapper.find(Home).find("div#avatar");

    expect(spy).toHaveBeenCalledTimes(1);
    expect(avatar.text()).toBe("KL");
  });
});
