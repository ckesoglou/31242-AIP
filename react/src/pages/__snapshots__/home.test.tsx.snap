// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render correctly 1`] = `
<Home
  history={Object {}}
  location={Object {}}
  match={Object {}}
/>
`;
